src/box/alter.cc

```diff
@@ -3387,6 +3419,19 @@ on_replace_dd_func(struct trigger * /* trigger */, void *event)
 				 "function has references");
 			return -1;
 		}
+
+		struct trigger *on_commit =
+			txn_alter_trigger_new(on_drop_func_commit, old_func);
+		struct trigger *on_rollback =
+			txn_alter_trigger_new(on_drop_func_rollback, old_func);
+		if (on_commit == NULL || on_rollback == NULL)
+			return -1;
+		txn_stmt_on_commit(stmt, on_commit);
+		txn_stmt_on_rollback(stmt, on_rollback);
+
+		/* Triggers pin the function - drop them before pin check. */
+		func_alter_triggers(old_func, false);
```

Comment:
Why do we need to pin the function by triggers if the triggers are deleted automatically when the function is dropped?

Reply:
It is needed to forbid deleting triggers that are being fired. The triggers are deleted from the event, which is actually unreferencing, not a deletion. So if the trigger is used at the moment, it won't be deleted, and the function will be still pinned - then the alter will fail.

Reply:
So if a trigger runs very often, it'll be quite hard to delete it?

Reply:
Yes, but "safety" is guaranteed then.
Also, if a trigger doesn't yield, the problem won't appear.